sdk/protocol2.go

```diff
@@ -0,0 +1,164 @@
+	sync.Mutex
+	Name               string        `json:"name"`
+	ProtocolVersion    string        `json:"protocol_version"`
+	IntegrationVersion string        `json:"integration_version"`
+	Data               []*EntityData `json:"data"`
+	prettyOutput       bool
+	writer             io.Writer
+}
+
+// NewIntegrationProtocol2WithWriter initializes a new instance of the integration protocol 2 specifying a writer instead of using stdout by default.
+func NewIntegrationProtocol2WithWriter(name string, version string, arguments interface{}, writer io.Writer) (*IntegrationProtocol2, error) {
+	i, err := NewIntegrationProtocol2(name, version, arguments)
+	if err != nil {
+		return nil, err
+	}
+
+	i.writer = writer
+
+	return i, nil
+}
+
+// NewIntegrationProtocol2 initializes a new instance of the integration protocol 2.
+func NewIntegrationProtocol2(name string, version string, arguments interface{}) (*IntegrationProtocol2, error) {
+	err := args.SetupArgs(arguments)
+	if err != nil {
+		return nil, err
+	}
+	defaultArgs := args.GetDefaultArgs(arguments)
+
+	log.SetupLogging(defaultArgs.Verbose)
+
+	// Avoid working with an uninitialized or in error state cache
+	if err = cache.Status(); err != nil {
+		return nil, err
+	}
+
+	integration := &IntegrationProtocol2{
+		Name:               name,
+		ProtocolVersion:    "2",
+		IntegrationVersion: version,
+		prettyOutput:       defaultArgs.Pretty,
+		writer:             os.Stdout, // StdOut by default
+	}
+
+	return integration, nil
+}
+
+// Entity method creates or retrieves an already created EntityData.
+func (integration *IntegrationProtocol2) Entity(entityName, entityType string) (*EntityData, error) {
+	if entityName == "" || entityType == "" {
+		return nil, errors.New("entityName and entityType should be not empty")
+	}
+
+	integration.Lock()
+	defer integration.Unlock()
+	for _, e := range integration.Data {
+		if e.Entity.Name == entityName && e.Entity.Type == entityType {
+			return e, nil
+		}
+	}
+
+	d := &EntityData{
+		Entity: Entity{
+			Name: entityName,
+			Type: entityType,
+		},
+	}
+
+	integration.Data = append(integration.Data, d)
+
+	return d, nil
+}
+
+// NewMetricSet returns a new instance of MetricSet with its sample attached to
+// the IntegrationData.
+func (d *EntityData) NewMetricSet(eventType string) metric.MetricSet {
+	ms := metric.NewMetricSet(eventType)
+	d.Metrics = append(d.Metrics, ms)
+
+	return ms
+}
+
+// Publish runs all necessary tasks before publishing the data. Currently, it
+// stores the cache, prints the JSON representation of the integration using a writer (stdout by default)
+// and re-initializes the integration object (allowing re-use it during the
+// execution of your code).
+func (integration *IntegrationProtocol2) Publish() error {
+	if err := cache.Save(); err != nil {
+		return err
+	}
+
+	output, err := integration.toJSON(integration.prettyOutput)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprint(integration.writer, output)
+	integration.Clear()
```

Comment:
if it returns error, doesn't have the integration to be cleared also?

Reply:
I don't think so. At this point the user should care about retrying if needed.